src/coreclr/jit/hwintrinsicarm64.cpp

```diff
@@ -482,31 +482,15 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
             // We shouldn't handle this as an intrinsic if the
             // respective ISAs have been disabled by the user.
 
-            if (sig->numArgs == 1)
-            {
-                op1     = impPopStack().val;
-                retNode = gtNewSimdHWIntrinsicNode(retType, op1, intrinsic, simdBaseJitType, simdSize);
-            }
-            else if (sig->numArgs == 2)
+            GenTree* args[32];
```

Comment:
I would use `ArrayStack<GenTree*>` or `GenTreeStack` instead (that is defined as `typedef ArrayStack<GenTree*> GenTreeStack`).

Reply:
@echesakovMSFT Hmm, this turns out more problematic that I would have liked.

- We only have one data structure with a preallocated size that allows us to avoid dynamic allocation in the common case, and that is `ArrayStack` (technically there is also a `SmallHashTable`).
- `ArrayStack` does not have direct ways to get a pointer to the underlying data. That's ok, we can add one.
- But then there is not a good pattern to use for popping the trees from the stack, as we must end up with an array that contains them in the order from last popped to first popped, while on the stack they're pushed in reverse order (the first to pop, the last in execution order, will end up on the bottom). So we actually need something like a queue, but a simple one won't do because we need a linear array at the end too.

In light of this, there are a couple solutions:
- Leave as-is.
- Expose the underlying array from `ArrayStack` and then use two loops, first to peek at the importer stack, second to pop the items:
```c++
ArrayStack<GenTree*> args(getAllocator(CMK_ImpStack)); // CMK_ImpStack - no better option, unfortunately
for (int i = sig->numArgs - 1; i >= 0; i--)
{
    args.Push(impStackTop(i).val);
}
for (int i = sig->numArgs - 1; i >= 0; i--)
{
    impStackPop();
}
```
- Leave a `TODO-List-Cleanup` for something(?).

What would your preference be (I do not have a strong opinion)?

Reply:
We could, yes, with supporting changes. I will see what it looks like in the end.

Reply:
@echesakovMSFT I have added support for moving the operands, PTAL...

To be honest I am not convinced the introduced complexity is worth it in this case. We were on the pretty safe side with the stack-allocated array of 256 bytes there, and adding wider vector support in the future will require expansive changes anyway.

Reply:
@SingleAccretion I agree - it looks more complex than needed. Let me take some time to play with the changes offline.

Reply:
@SingleAccretion I took a stab at simplifying the logic...

TBH, I didn't fully understand the need for `IntrinsicNodeBuilder` at the first place. However, after couple attempts of trying to make `GenTreeSIMD`/`GenTreeHWIntrinsic` constructors to re-use `gtInlineOperands` instead of allocating an array - I realized how you ended up with the solution.

One idea that I had was moving the logic corresponding to allocation of an array/re-using `gtInlineOperands` from `IntrinsicNodeBuilder` to `GenTreeJitIntrinsic` constructor and introducing three overloads of the constructor for 0, 1 and 2 operands - that would pass `gtInlineOperands`, `operandCount` down to `GenTreeMultiOp` as storage/size. Although this has a minor benefit of not requiring to pass `CompAllocator`/ `IntrinsicNodeBuilder`, I don't believe the implementation is any simpler than what you have.

At this point, I am quite convinced that the complexity is justified.